### PR TITLE
Chooser scenario

### PR DESCRIPTION
**Pull request by fgmccabe**
This is an initial draft.
There are some variations on instructions and some instruction renaming.

### CODE REVIEW

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 1]
```diff
@@ -0,0 +1,680 @@
...
+  local.get $str
+  string.size
+  call-export "malloc-x"     ;; create string entity of right size
+  own (i32)
+    call-export "free-x"     ;; temporary, until safely re-owned
+  end
+  let (local $ptr owned<i32>) (result owned<i32>)
+    local.get $str
+    local.get $ptr
+	owned.access             ;; access the owned pointer
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: is this indented deliberately? It seems like, but I don't understand the pattern.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+
+
+```wasm
+(@interface func $stralloc-x
+  (param $str string)
+  (result owned<i32>)        ;; return an owned shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-x"     ;; create string entity of right size
+  own (i32)
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: Is `own` described somewhere? It looks block-like, but it appears the code below is actually a destructor.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 3]
```diff
@@ -0,0 +1,680 @@
...
+    call-export "free-x"     ;; temporary, until safely re-owned
+  end
+  let (local $ptr owned<i32>) (result owned<i32>)
+    local.get $str
+    local.get $ptr
+	owned.access             ;; access the owned pointer
+	local.get $str
+	string.size
+    utf8.from.string
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: I found this to be a bit hard to follow, might be useful to have a stack diagram too (or use the folded format)

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 4]
```diff
@@ -0,0 +1,680 @@
...
+
+In addition to the Interface Types memory management, because we are using a
+non-trivial C++ structure, we have to invoke the appropriate constructors,
+access functions and destructors of our `shared_ptr` structure.
+
+This is achieved through some additional core helper functions that are exported
+by the core wasm module, although they would not typically be exported as
+Interface Type functions:
+
+* `shared-builder-x` is used to create and initialize a `shared_ptr`
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: include func signature for these?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 5]
```diff
@@ -0,0 +1,680 @@
...
+
+(@interface func (export "chooser")
+  (param $left string)
+  (param $right string)
+  (result string)
+
+  local.get $left
+  invoke-func $stralloc-x
+  invoke-func $shared-x     ;; make it a shared ptr
+  let (local $lpo owned<i32>) (result string)
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: minor nit: current function-references proposal requires this the other way around:
```
let (result string) (local $lpo owned<i32>)
```

TBH, I think it's more clear the way you have it, but I thought I'd mention.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+* `shared-builder-x` is used to create and initialize a `shared_ptr`
+  structure. It takes a pointer as an argument and returns a `shared_ptr` with
+  an initial reference count of 1.
+
+* `shared-release-x` is used to decrement the reference count of a `shared_ptr`;
+  and if that results in the reference count dropping to zero _both the
+  underlying resource and the `shared_ptr` structure itself are disposed of_.
+
+* `shared-get-x` is used to access the underlying resource of a `shared_ptr`
+  without affecting it's reference count.
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: nit: its

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 7]
```diff
@@ -0,0 +1,680 @@
...
+The import adapter for `chooser_` has to lift the two argument `string`s and
+lower the return value:
+
+```wasm
+(@interface func $stralloc-i
+  (param $str string)
+  (result i32)           ;; return a shared ptr of the string
+  local.get $str
+  string.size
+  call-export "malloc-i"     ;; create string entity of right size
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: This allocates the string, but not the shared_ptr that wraps it, right?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+      end
+	end	
+  end
+  invoke-func "stralloc-i" ;; allocate a local string for the result
+)
+```
+
+After we expand the helper functions, we get:
+
+```wasm2
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: Seems like only `wasm` works for giving syntax highlighting

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  local.get $r
+  call-export "access-utf8-i"
+  string.from.utf8
+
+  let ($left string
+       $right string)
+    local.get $left
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: It'd be nice to bracket these inlined functions somehow, maybe just with `;; invoke-func $str-alloc-x`, etc.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 10]
```diff
@@ -0,0 +1,680 @@
...
+      call-export "shared-builder-x" ;; make it a shared ptr pair
+      local.tee $o1
+    end
+    let (local $lp i32) (result i32)
+      local.get $r
+      call-export "access-utf8-i"
+      let (local $rbase i32)(local $rsize i32)(result i32)
+        local.get $rbase
+        local.get $rsize
+        call-export "malloc-x"
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: it looks like we lost the matching `free-x` here...?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 11]
```diff
@@ -0,0 +1,680 @@
...
+
+For simplicity, we migrated all the deallocations to the end of the fused
+adapter. In some situations, for example when processing arrays, we may wish to
+be more aggressive in invoking the memory release code.
+
+Although fairly long, this fused adapter has a striaghtfoward structure: the
+input strings are copied from the import memory to the export memory -- and also
+wrapped as shared pointer structures as required by the signature of the
+`nondeterministic_chooser` function. The resulting string is copied from the
+export localtion to the import memory. And, finally, any temporary structures
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: sp: location

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+allocated are released.
+
+Note that the string returned by `nondeterministic_chooser` is not directly
```
Comment (binji) [inline proposals/interface-types/working-notes/scenarios/chooser.md]: I think I understand what this is getting at, but the sentence doesn't parse for me

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 13]
```diff
@@ -0,0 +1,680 @@
...
+
+This results in _two_ memory allocations for the shared resource: one for the
+resource itself and one for the pointer structure -- which contains a reference
+count and a raw pointer to the resource.
+
+
+```wasm
+(@interface func $stralloc-x
+  (param $str string)
+  (result owned<i32>)        ;; return an owned shared ptr of the string
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:74]: Nit: I know we've been writing `owned<i32>` a lot informally, but I think in wat syntax this should be `(owned i32)`.  The reason is that `owned` is a type constructor (symmetric to `func`, `struct`, `array`, etc) and `i32` is its argument and wasm writes type-constructors-applied-to-arguments as `(type-ctor args...)`.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 1 above)
Comment (lukewagner) [reply proposals/interface-types/working-notes/scenarios/chooser.md:84]: FWIW, +1.  I think a better scheme for handling when the linear form becomes obtuse is switching to the [folded](https://webassembly.github.io/spec/core/text/instructions.html#text-foldedinstr) form.  Notice that you can fold only just the parts you want and otherwise keep things in sequence.  Otherwise, I wouldn't invent a custom indention scheme.  (FWIW, in email threads, I've been wondering whether the indention was an unintended email artifact :P)

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 5 above)
Comment (lukewagner) [reply proposals/interface-types/working-notes/scenarios/chooser.md:114]: Agreed it reads more naturally to put results second, making `let` more like a function signature where `param`s have been replaced by `local`s.  Maybe we should change the function-references proposal...

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 14]
```diff
@@ -0,0 +1,680 @@
...
+    local.get $ptr
+	owned.access             ;; access the owned pointer
+	local.get $str
+	string.size
+    utf8.from.string
+	local.get $ptr
+  end
+)
+
+(@interface func $shared-x
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:92]: nit: maybe name `sharedalloc-x` for symmetry with `stralloc-x`?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 15]
```diff
@@ -0,0 +1,680 @@
...
+      local.get $lp
+	  owned.access
+      local.get $rp
+	  owned.access
+      call-export "nondeterministic_choice"  ;; call the chooser itself
+      own (i32)               ;; own the result
+        call-export "shared-release-x"    ;; will eventually call to free string
+      end		
+      owned.access
+	  call-export "shared-get-x"
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:129]: nit: maybe name `shared-get-ptr-x`?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 16]
```diff
@@ -0,0 +1,680 @@
...
+>part of the adapter fusion process.
+
+The returned value from `nondeterministic_choice` is wrapped up as an `own`ed
+allocation and lifted to a `string`.
+
+>Note we do not need to `own` the string memory of the arguments because we have
+>asserted that both the arguments to `nondeterministic_choice` will be the
+>_last_ references to the strings. However, only one of the C++ strings will be
+>deallocated -- the other is returned to us. We _do_ need to `own` the return
+>result, however.
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:154]: I think the ownership of the arguments and results can/should be considered orthogonally: the code as written works equally well if `nondeterministic_choice` returned a `shared_ptr<std::string>(new std::string("foo"))`, instead of one of its parameters.  In particular, the clause "because we have asserted that both the arguments to `nondeterministic_choice` will be the last references to the strings" should not be necessary: an equally-valid ABI is that the calling adapter code performs `owned.release` on arguments right before calling `nondeterministic_choice`, thereby handing off responsibility to decrement to the caller instead of having the caller *independently* increment the given pointer.  This would be slightly more efficient (less increments/decrements) and more symmetric to how returns work, so could we do that instead?

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 17]
```diff
@@ -0,0 +1,680 @@
...
+  (result i32))
+```
+
+The two strings are passed as memory addresses of structures -- from which the
+text of the string and its length can be ascertained. The structure itself is
+unspecified; we will use access functions -- such as `access-utf8` as needed.
+
+>Note that although _we_ believe that the returned value will be the same as one
+>of the arguments, the limitations of Interface Types mean that the returned
+>string will be a copy of one of the arguments.
```
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:199]: "we" in this sentence is a bit ambiguous; maybe say "Note that, although *ideally* `chooser_` would simply return one of its parameters, the limitations of ..."

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 7 above)
Comment (lukewagner) [inline proposals/interface-types/working-notes/scenarios/chooser.md:210]: I think you should `own` the result of `malloc-i` with a call to free, making the return type `(owned i32)`, and adding an `owned.release` right before returning in `chooser_`.  Otherwise any exception between here and the final return would leak this memory.

**Comment by lukewagner** [comment]
Also: great job!  This was a lot of work.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 1 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:84]: That particular indentation was an unexpected consequence of using emacs :)
I may try the folded form in a future revision

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 2 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:78]: To be documented.
It is a block, that is 'attached' to the value; to be executed when we are done with the memory.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 3 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:87]: will attempt

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 4 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:164]: good idea. But, the interesting signature is neither wasm nor c++ nor interface types.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 6 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:173]: My excuse: English is my second language :)

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 7 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:210]: It allocates the buffer in which to put the string text in.
The shared_ptr is a separate allocation

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 13 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:74]: I would like to follow this up in a different conversation.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 10 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:618]: the own.release cancels the own block.
if we were accounting for exceptions properly (which we are not yet) then the free-x call would should up in the handler.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 18]
```diff
@@ -0,0 +1,680 @@
...
+non-determinism.
+
+This scenario is based on the idea of using run-time information in order to
+control the flow of information. In particular, consider the C++ chooser
+function:
+
+```C++
+typedef std::shared_ptr<std::string> shared_string;
+
+shared_string nondeterministic_choice(shared_string one, shared_string two) {
```
Comment (jgravelle-google) [inline proposals/interface-types/working-notes/scenarios/chooser.md:20]: This is an odd C++ function. std::move mixed with shared_ptr is a weird one. It makes it unclear whether we're trying to re-use the existing shared_ptr without incrementing the refcounts, and if so it's unclear why we're using shared_ptr instead of directly passing a `std::string&&`.

This sample feels overloaded, like it's trying to illustrate three things at once. At least it needs to solve three things:

1. how to handle C++ shared_ptrs
2. how to reuse a passed-in reference (simpler version: identity function)
3. how to handle conditional ownership

It would be easier to follow and discuss if these were more explicitly separated. As it is now, they seem subtly interleaved, and it's hard to disentangle problems with shared_ptr vs problems with ownership.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 1 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:84]: parenthetical note: the folded form relates to expression evaluation.
I was using indentation to reflect scope structure and other block structuring; not the same

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 18 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:20]: The reason for choosing the chooser is to highlight the necessity of the machinery. We start with a non-deterministic scenario and show how to implement it.
However, I defer to others wrt C++ style. What I was trying to do was build a scenario where the input may come back (or not) in the return result, and where copying would not be an acceptable answer.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 5 above)
Comment (binji) [reply proposals/interface-types/working-notes/scenarios/chooser.md:114]: Ah, but there's a subtlety to `let` that I had missed. `let` introduces a new block, so it has a blocktype, which means w/ the multi-value proposal it's possible to have params and locals:

```wasm
f32.const 1
i32.const 2
let (param i32) (result i64) (local f32)
  ;; i32:2 is still on top of the stack
  ;; f32:1 was consumed from the stack, and is now local 0
  ...
end
```

I'm not sure there's value in having `let` params, but it does feel a bit odd to have the behavior differ significantly from the way other blocks work. It also means that a `let` block doesn't work at all like the locals at the beginning of a function (which means function inlining doesn't work naturally either).

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 8 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md]: that was not intended for publication. Now removed.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 5 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md]: I missed that too. May need to revisit.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 9 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md]: I have taken out a lot of the intermediate stages. Will write this process up separately.  The final code is not always immediately recognizable anymore - so labeling is problematic.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 12 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:385]: reworded some. Missing text or over-enthusiastic use of delete ...

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 16 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:144]: dropped some of the text here.

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+core WASM.
+
+In this scenario we take a somewhat simplified view of C++'s implementation of
+shared pointers: a shared pointer is implemented as a pair consisting of a
+reference count and a raw pointer to the shared resource.
+
+>Note: In practice, the C++ implementation of shared pointers is somewhat more
+>complex; for reasons that are not important here.
+
+This results in _two_ memory allocations for the shared resource: one for the
```
Comment (ajklein) [inline proposals/interface-types/working-notes/scenarios/chooser.md:70]: It's probably a little late for this (since I think this code does work), but did you consider passing the shared pointers as pairs instead of allocated structs? That would have reduced the number of allocations in the example, and (to my reading) made it more "realistic" by passing the refcounts by value instead of by reference (as is implicit in the example code).

File: proposals/interface-types/working-notes/Instructions.md [hunk 20]
```diff
@@ -1,292 +0,0 @@
-# Interface instructions
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply proposals/interface-types/working-notes/Instructions.md:1]: I think that this is because these files were accidentally included once. I deleted them in a subsequent commit but github fu is hard to figure out

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 21]
```diff
@@ -0,0 +1,402 @@
...
+  end
+)
+
+(@interface func $sharedalloc-x
+  (param $ptr (owned i32))
+  (result (owned i32))        ;; return a shared ptr of the entity
+
+  local.get $ptr
+  owned.release              ;; access pointer and remove ownership
+  i32.const 1                ;; initial reference count of 1
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:101]: good catch!

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 22]
```diff
@@ -0,0 +1,402 @@
...
+  (param $right string)
+  (result string)
+
+  (call-export "nondeterministic_choice"  ;; call the chooser itself
+    (invoke-func $sharedalloc-x     ;; make it a shared ptr
+      (invoke-func $stralloc-x
+        (local.get $left)))
+    (invoke-func $sharedalloc-x   ;; make it a shared ptr
+      (invoke-func $stralloc-x
+        local.get $right)))
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:120]: thx

File: proposals/interface-types/working-notes/scenarios/chooser.md [hunk 23]
```diff
@@ -0,0 +1,402 @@
...
+that is that one cannot easily predict when given allocated blocks may be freed.
+
+Our exemplar for this is the non-deterministic C++ chooser function; which
+(might) return one of its `string` arguments as its return value:
+
+```C++
+typedef std::shared_ptr<std::string> shared_string;
+
+shared_string nondeterministic_choice(shared_string one, shared_string two) {
+  return random() > 0.5 ? std::move(one) : std_move(two);
```
[the remark this answers is not in the document]
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:24]: fixed the automatically generated typo. But, I was assured that the std::move is necessary.

File: proposals/interface-types/working-notes/scenarios/chooser.md (hunk 19 above)
Comment (fgmccabe) [reply proposals/interface-types/working-notes/scenarios/chooser.md:70]: Ah. Was trying to emulate a simplified version of shared_ptr. Cant pass refcounts themselves by value; (suppose callee decides to take an extra reference to a string?)

Comment (ajklein) [reply proposals/interface-types/working-notes/scenarios/chooser.md:70]: When I woke up this morning I'd realized where I went awry: while "real" shared_ptrs themselves are passed-by-value, they share an underlying auxiliary structure. Whereas in your case the shared_ptr itself is that structure. That makes the ABI for the function rather surprising for something with passed-by-value parameters, but in the end it amounts to the same thing.
